--- wget-1.17/src/hsts.c.orig	Wed Nov 25 03:41:10 2015
+++ wget-1.17/src/hsts.c	Wed Nov 25 03:48:18 2015
@@ -49,6 +49,7 @@
 #include <string.h>
 #include <stdio.h>
 #include <sys/file.h>
+#include <fcntl.h>
 
 struct hsts_store {
   struct hash_table *table;
@@ -494,6 +495,7 @@
   struct stat st;
   FILE *fp = NULL;
   int fd = 0;
+  struct flock lock;
 
   if (filename && hash_table_count (store->table) > 0)
     {
@@ -502,7 +504,11 @@
         {
           /* Lock the file to avoid potential race conditions */
           fd = fileno (fp);
-          flock (fd, LOCK_EX);
+          lock.l_start = 0;
+          lock.l_len = 0;
+          lock.l_type = F_WRLCK;
+          lock.l_whence = SEEK_SET;
+          fcntl(fd, F_SETLKW, &lock);
 
           /* If the file has changed, merge the changes with our in-memory data
              before dumping them to the file.
