--- wget-1.17/doc/Makefile.in.orig	Wed Nov 25 04:09:36 2015
+++ wget-1.17/doc/Makefile.in	Wed Nov 25 04:09:48 2015
@@ -1792,7 +1792,7 @@
 	$(TEXI2POD) -D VERSION="$(VERSION)" $(srcdir)/wget.texi $@
 
 $(MAN): wget.pod
-	$(POD2MAN) --center="GNU Wget" --release="GNU Wget @VERSION@" --utf8 $? > $@
+	$(POD2MAN) --center="GNU Wget" --release="GNU Wget @VERSION@" $? > $@
 
 #wget.cat: $(MAN)
 #	nroff -man $? > $@
