--- include/curl/Makefile.in.orig	Tue Feb 24 15:05:38 2015
+++ include/curl/Makefile.in	Thu Mar 12 00:10:50 2015
@@ -439,13 +439,9 @@
 	cd $(top_builddir) && $(MAKE) $(AM_MAKEFLAGS) am--refresh
 $(am__aclocal_m4_deps):
 
-curlbuild.h: stamp-h2
-	@test -f $@ || rm -f stamp-h2
-	@test -f $@ || $(MAKE) $(AM_MAKEFLAGS) stamp-h2
-
-stamp-h2: $(srcdir)/curlbuild.h.in $(top_builddir)/config.status
+stamp-h2:
 	@rm -f stamp-h2
-	cd $(top_builddir) && $(SHELL) ./config.status include/curl/curlbuild.h
+	touch stamp-h2
 
 distclean-hdr:
 	-rm -f curlbuild.h stamp-h2
