--- bind-9.6-ESV-R10-P2/lib/isc/unix/ifiter_getifaddrs.c.orig	Tue Oct 13 06:46:56 2015
+++ bind-9.6-ESV-R10-P2/lib/isc/unix/ifiter_getifaddrs.c	Tue Oct 13 07:03:03 2015
@@ -143,7 +143,7 @@
 	if (ifa->ifa_addr == NULL)
 		return (ISC_R_IGNORE);
 
-	family = ifa->ifa_addr->sa_family;
+	family = ifa->ifa_addr->ss_family;
 	if (family != AF_INET && family != AF_INET6)
 		return (ISC_R_IGNORE);
 
@@ -174,15 +174,18 @@
 
 	iter->current.af = family;
 
-	get_addr(family, &iter->current.address, ifa->ifa_addr, ifa->ifa_name);
+	get_addr(family, &iter->current.address,
+		 (struct sockaddr *)(ifa->ifa_addr), ifa->ifa_name);
 
 	if (ifa->ifa_netmask != NULL)
-		get_addr(family, &iter->current.netmask, ifa->ifa_netmask,
+		get_addr(family, &iter->current.netmask,
+			 (struct sockaddr *)(ifa->ifa_netmask),
 			 ifa->ifa_name);
 
 	if (ifa->ifa_dstaddr != NULL &&
 	    (iter->current.flags & INTERFACE_F_POINTTOPOINT) != 0)
-		get_addr(family, &iter->current.dstaddress, ifa->ifa_dstaddr,
+		get_addr(family, &iter->current.dstaddress,
+			 (struct sockaddr *)(ifa->ifa_dstaddr),
 			 ifa->ifa_name);
 
 	return (ISC_R_SUCCESS);
