--- src/tspi/obj_rsakey.c.orig	Fri Oct 16 22:24:34 2015
+++ src/tspi/obj_rsakey.c	Fri Oct 16 22:25:28 2015
@@ -1721,7 +1721,7 @@
 TSS_RESULT
 obj_rsakey_set_srk_pubkey(BYTE *pubkey)
 {
-	struct tsp_object *obj, *prev = NULL;
+	struct tsp_object *obj;
 	struct obj_list *list = &rsakey_list;
 	struct tr_rsakey_obj *rsakey;
 	TSS_RESULT result;
@@ -1728,7 +1728,7 @@
 
 	MUTEX_LOCK(list->lock);
 
-	for (obj = list->head; obj; prev = obj, obj = obj->next) {
+	for (obj = list->head; obj; obj = obj->next) {
 		rsakey = (struct tr_rsakey_obj *)obj->data;
 
 		/* we found the SRK, set this data as its public key */
@@ -1895,13 +1895,13 @@
 void
 obj_rsakey_remove_policy_refs(TSS_HPOLICY hPolicy, TSS_HCONTEXT tspContext)
 {
-	struct tsp_object *obj, *prev = NULL;
+	struct tsp_object *obj;
 	struct obj_list *list = &rsakey_list;
 	struct tr_rsakey_obj *rsakey;
 
 	MUTEX_LOCK(list->lock);
 
-	for (obj = list->head; obj; prev = obj, obj = obj->next) {
+	for (obj = list->head; obj; obj = obj->next) {
 		if (obj->tspContext != tspContext)
 			continue;
 
