--- Python-2.6.9/configure.o1	Mon Oct 19 00:09:43 2015
+++ Python-2.6.9/configure	Mon Oct 19 00:11:39 2015
@@ -6154,71 +6154,10 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $was_it_defined" >&5
 $as_echo "$was_it_defined" >&6; }
 
-# Check whether using makedev requires defining _OSF_SOURCE
-{ $as_echo "$as_me:${as_lineno-$LINENO}: checking for makedev" >&5
-$as_echo_n "checking for makedev... " >&6; }
-cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
-
-#if defined(MAJOR_IN_MKDEV)
-#include <sys/mkdev.h>
-#elif defined(MAJOR_IN_SYSMACROS)
-#include <sys/sysmacros.h>
-#else
-#include <sys/types.h>
-#endif
-int
-main ()
-{
- makedev(0, 0)
-  ;
-  return 0;
-}
-_ACEOF
-if ac_fn_c_try_link "$LINENO"; then :
-  ac_cv_has_makedev=yes
-else
-  ac_cv_has_makedev=no
-fi
-rm -f core conftest.err conftest.$ac_objext \
-    conftest$ac_exeext conftest.$ac_ext
-if test "$ac_cv_has_makedev" = "no"; then
-    # we didn't link, try if _OSF_SOURCE will allow us to link
-    cat confdefs.h - <<_ACEOF >conftest.$ac_ext
-/* end confdefs.h.  */
-
-#define _OSF_SOURCE 1
-#include <sys/types.h>
-
-int
-main ()
-{
- makedev(0, 0)
-  ;
-  return 0;
-}
-_ACEOF
-if ac_fn_c_try_link "$LINENO"; then :
-  ac_cv_has_makedev=yes
-else
-  ac_cv_has_makedev=no
-fi
-rm -f core conftest.err conftest.$ac_objext \
-    conftest$ac_exeext conftest.$ac_ext
-    if test "$ac_cv_has_makedev" = "yes"; then
-
-$as_echo "#define _OSF_SOURCE 1" >>confdefs.h
-
-    fi
-fi
-{ $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_has_makedev" >&5
-$as_echo "$ac_cv_has_makedev" >&6; }
-if test "$ac_cv_has_makedev" = "yes"; then
-
+# Solaris has makedev, and posixmodule already has all the right #include
+# statements.
 $as_echo "#define HAVE_MAKEDEV 1" >>confdefs.h
 
-fi
-
 # Enabling LFS on Solaris (2.6 to 9) with gcc 2.95 triggers a bug in
 # the system headers: If _XOPEN_SOURCE and _LARGEFILE_SOURCE are
 # defined, but the compiler does not support pragma redefine_extname,
